api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -379,16 +414,31 @@ def __init__(
                 The number of rows of ``obs`` and ``X`` data to return in each iteration. Defaults to 1. A value of 1
                 will result in Tensors of rank 1 being returns (a single row); larger values will result in Tensors of
                 rank 2 (multiple rows).
+            shuffle:
+                Whether to shuffle the ``obs`` and ``X`` data being returned. Defaults to ``False`` (no shuffling).
+                For performance reasons, shuffling is performed in two steps: 1) a global shuffling of contiguous
+                ranges of data, where the rows are grouped into chunks and
+                the chunks themselves are shuffled, and 2) a local shuffling, where each chunk is shuffled internally.
+                The global shuffling ensures that the ordering of result data will be randomized across the entire obs
+                axis, while also reducing the number of SOMA read operations by leveraging the spatial locality of
+                contiguously stored data. The local shuffling ensures the data is shuffled at the row-level within each
+                chunk.
+            seed:
+                The random seed used for shuffling. Defaults to ``None`` (no seed). This *must* be specified when using
+                ``DistributedDataParallel`` to ensure data partitions are disjoint across worker processes.
             return_sparse_X:
                 Controls whether the ``X`` data is returned as a dense or sparse Tensor. As ``X`` data is very sparse,
                 setting this to ``True`` will reduce memory usage, if the model supports use of sparse Tensors. Defaults
                 to ``False``, since sparse Tensors are still experimental in PyTorch.
-            soma_buffer_bytes:
-                The number of bytes to use for reading data from SOMA. If not specified, will use the default SOMA
-                value. Maximum memory utilization is controlled by this parameter, with larger values providing better
-                read performance.
+            soma_chunk_size:
+                The number of obs/X rows to retrieve when reading data from SOMA. This impacts two aspects of
+                ``ExperimentDataPipe`` behavior: 1) The maximum memory utilization is controlled by this
+                parameter, with larger values providing better read performance, but also requiring more memory. If
+                not specified, the value is set to utilize ~1 GiB of RAM per SOMA chunk read, based upon the number
+                of ``var`` columns (cells/features) being requested and assuming X data sparsity of 95%.
+                2) The granularity of the global shuffling step (see ``shuffle`` parameter).
```

Comment:
I wonder if it's worth providing some extra guidance similar to what you did in 1) e.g. "Large values lead to diminished global shuffling, and small values lead to increased global shuffling"

Reply:
I'm going with:

>  A large chunk size will result in fewer chunks, resulting in an overall less uniform distribution (unless chunk count is 1); conversely, smaller chunk sizes will result in a larger number of chunks, which will result in an overall more uniform distribution.

This is a statistically interesting problem, in fact, and it's not immediately clear to me how _exactly_ the distribution uniformity changes as function of chunk size, or inversely, chunk count. When either reaches 1, a uniform distribution should result, as a single chunk or single-rowed chunks are both equivalent to a one-step shuffling process, where all the rows are sorted at once. However my LLM-based sources indicate the above is true; but I haven't found any directly-related literature on the matter. I'll have to start running some Monte Carlo simulations...

Reply:
I believe we are aligned, see my comments below.

I think speaking of uniformity may confuse the reader. I would suggest a slight change in language

> A large chunk size will result in fewer chunks, resulting in reduced shuffling across all cells in the query; conversely, smaller chunk sizes will result in a larger number of chunks, which will result in an increased shuffling across all cells in the query.

------

I think in isolation your global shuffler without TorchData’s shuffler, would do effectively no shuffling (or very minimal) for training if the “SOMA chunk size” >> “Training batch size”

So the power and complexity comes when mixing your global shuffler with the TorchData’s shuffler. But the amount of true randomness, i.e. how close we get to effectively have shuffling that represents random ordering of all individual cells would be defined by the following variables:

1. Batch modeling size
2. SOMA chunk size.

So fixing all other variables except SOMA chunk size:

1. Global shuffler is enabled
2. TorchData’s shuffler is enabled
3. Batch modeling size is fixed at X value (e.g. 128)

Then the smaller the SOMA chunk size the closer we get to complete random reshuffling of all cells, with the extremes SOMA chunk size of 1 leading to full reshuffled representation and SOMA chunk size of N (where N is the total number of cells) leads the lowest level of shuffling (i.e. only shuffling within batch size)

I think it may good to get our theory straight, and do some simulations if we are still not sure about it

Reply:
In fact, there is no need for the TorchData shuffler now.  ExperimentDataPipe's global/local shuffling ensures that every cell gets shuffled into a new position for the entire query result.  Torch batches are drawn from the fully shuffled result.  The soma_chunk_size only determines how far apart certain cells can end up from one another in the shuffled result. Cells within a given chunk can never be farther than the chunk size.

This might require further discussion in person to make sure we're aligned.